socorro/app/fetch_transform_save_app.py

```diff
@@ -104,32 +110,140 @@ def get_application_defaults():
+            self.source_iterator = self._infinite_iterator
+        elif config.number_of_submissions == 'all':
+            self.source_iterator = self._all_iterator
+        else:
+            self.source_iterator = self._limited_iterator
+
+    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+    # iterator section
+    #
+    # the following methods setup a nested hierarchy of iterators.  Each layer
+    # in the hierarchy adds a feature to the enclosed iterator.
+
+    # The  iterator at the core, the innermost, will yield a list of crash_ids
+    # from some source. A subclass may define its own source and makes it
+    # available by overriding the method "_create_iter".
+    #
+    # That innermost iterator is then wrapped by the "_basic_iterator" method.
+    # this iterator changes the form of the yielded values to the (*args,
+    # **kwargs) as required by the TaskManager.  It also gives derived classes
+    # the oportunity to run callbacks between each iteration or at the point
+    # that the innermost iterator is exhausted.
+    #
+    # The outermost layer of the iterator nesting imposes length limitations
+    # on the iterators.  Controlled by the configuration parameter
+    # "number_of_submissions" this iterator can be limit to an exact number
+    # of iterations, set to run forever, or run only until the natural
+    # exhaustion of the innermost iterator.
+    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+    #--------------------------------------------------------------------------
+    def _create_iter(self):
+        # the actual mechanism of creating the iterator to be overridden in
+        # subclasses based on what they want to iterate over.  In this default,
+        # iteration will come from the crashstorage instance defined as
+        # "source".  This works for the crashmover where the generally used
+        # source is the file system walking class.
+        return self.source.new_crashes()
+
+    #--------------------------------------------------------------------------
+    def _action_between_each_iteration(self):
+        """an action to take after an item has been yielded by the iterator.
+        This method is to be overridden by derived classes"""
+        pass
 
     #--------------------------------------------------------------------------
-    def source_iterator(self):
-        """this iterator yields individual crash_ids from the source
-        crashstorage class's 'new_crashes' method."""
-        while(True):  # loop forever and never raise StopIteration
-            for x in self.source.new_crashes():
-                if x is None:
-                    yield None
-                elif isinstance(x, tuple):
-                    yield x  # already in (args, kwargs) form
-                else:
-                    yield ((x,), {})  # (args, kwargs)
+    def _action_after_iteration_completes(self):
```

Comment:
Perhaps, again, this is a matter of style/taste but this looks like it gives an implementor a chance to override or plug in to the API. But the method is named as a private method which tells the implementor to "stay away".